driver/src/main/com/mongodb/MongoDatabaseImpl.java

```diff
@@ -132,25 +131,71 @@ public Document runCommand(final Bson command, final ReadPreference readPreferen
+        return executeCommand(null, command, readPreference, resultClass);
+    }
+
+    @Override
+    public Document runCommand(final ClientSession clientSession, final Bson command) {
+        return runCommand(clientSession, command, ReadPreference.primary(), Document.class);
+    }
+
+    @Override
+    public Document runCommand(final ClientSession clientSession, final Bson command, final ReadPreference readPreference) {
+        return runCommand(clientSession, command, readPreference, Document.class);
+    }
+
+    @Override
+    public <TResult> TResult runCommand(final ClientSession clientSession, final Bson command, final Class<TResult> resultClass) {
+        return runCommand(clientSession, command, ReadPreference.primary(), resultClass);
+    }
+
+    @Override
+    public <TResult> TResult runCommand(final ClientSession clientSession, final Bson command, final ReadPreference readPreference,
+                                        final Class<TResult> resultClass) {
+        notNull("clientSession", clientSession);
+        return executeCommand(clientSession, command, readPreference, resultClass);
+    }
+
+    private <TResult> TResult executeCommand(final ClientSession clientSession, final Bson command, final ReadPreference readPreference,
+                                             final Class<TResult> resultClass) {
         notNull("readPreference", readPreference);
         return executor.execute(new CommandReadOperation<TResult>(getName(), toBsonDocument(command), codecRegistry.get(resultClass)),
-                readPreference);
+                readPreference, clientSession);
     }
 
     @Override
     public void drop() {
-        executor.execute(new DropDatabaseOperation(name, getWriteConcern()));
+        executeDrop(null);
+    }
+
+    @Override
+    public void drop(final ClientSession clientSession) {
+        executeDrop(clientSession);
+    }
+
+    private void executeDrop(final ClientSession clientSession) {
+        executor.execute(new DropDatabaseOperation(name, getWriteConcern()), clientSession);
     }
 
     @Override
     public MongoIterable<String> listCollectionNames() {
-        return new ListCollectionsIterableImpl<BsonDocument>(name, BsonDocument.class, getDefaultCodecRegistry(), ReadPreference.primary(),
-                executor).map(new Function<BsonDocument, String>() {
-            @Override
-            public String apply(final BsonDocument result) {
-                return result.getString("name").getValue();
-            }
-        });
+        return executeListCollectionNames(null);
+    }
+
+    @Override
+    public MongoIterable<String> listCollectionNames(final ClientSession clientSession) {
+        return executeListCollectionNames(clientSession);
+    }
+
+    private MongoIterable<String> executeListCollectionNames(final ClientSession clientSession) {
+        notNull("clientSession", clientSession);
```

Comment:
Can't have the check here - as the `listCollectionNames` will call this with a null.

Reply:
Same as above.  This method is not tested.